File: bindings/flutter/lib/src/foundry_local.dart
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import 'dart:async';
import 'dart:convert';
import 'dart:ffi';

import 'package:ffi/ffi.dart';
import 'package:meta/meta.dart';
import 'package:path_provider/path_provider.dart' as pp;

import 'bindings/bindings.dart' as raw;
import 'bindings/native_library.dart';
import 'error_capture.dart';
import 'job_runner.dart';
import 'lifecycle.dart';
import 'model.dart';
import 'models/config.dart';
import 'models/device_profile.dart';
import 'models/errors.dart';
import 'models/progress.dart';
import 'native_strings.dart';

/// Root object of the SDK. Owns the manager handle and the app-lifecycle
/// bridge; released by [dispose].
class FoundryLocal {
  FoundryLocal._({
    required this.config,
    required int managerHandle,
    required LifecycleBridge lifecycleBridge,
  })  : _managerHandle = managerHandle,
        _lifecycleBridge = lifecycleBridge;

  /// Configuration this instance was created with. Read-only after creation.
  final FoundryLocalConfig config;

  final int _managerHandle;
  final LifecycleBridge _lifecycleBridge;
  bool _disposed = false;

  /// Runtime version reported by the loaded core, or null when the runtime
  /// shared library is missing.
  static String? get runtimeVersion {
    final ptr = NativeLibrary.instance.bindings.flm_runtime_version_string();
    if (ptr == nullptr) return null;
    final s = cStringToDart(ptr);
    return s.isEmpty ? null : s;
  }

  /// SDK version string (matches the plugin's pubspec version).
  static String get sdkVersion =>
      cStringToDart(NativeLibrary.instance.bindings.flm_version_string());

  /// Whether the ONNX Runtime GenAI runtime is present and loadable.
  static bool get isRuntimeAvailable =>
      NativeLibrary.instance.bindings.flm_is_runtime_available() != 0;

  /// Create a manager.
  ///
  /// The plugin fills in `app_data_dir` from `getApplicationSupportDirectory`
  /// unless the caller sets it explicitly, because the core requires a sandbox
  /// path on mobile.
  static Future<FoundryLocal> create(FoundryLocalConfig config) async {
    final bindings = NativeLibrary.instance.bindings;

    var resolvedConfig = config;
    if (config.appDataDir == null) {
      final dir = await pp.getApplicationSupportDirectory();
      resolvedConfig = FoundryLocalConfig(
        appName: config.appName,
        appDataDir: dir.path,
        logLevel: config.logLevel,
        autoUnloadOnBackground: config.autoUnloadOnBackground,
        jobPoolThreads: config.jobPoolThreads,
        additionalOptions: config.additionalOptions,
      );
    }

    late int managerHandle;
    withCString(resolvedConfig.toJsonString(), (cfgPtr) {
      final out = calloc<Uint64>();
      try {
        final status = bindings.flm_manager_create(cfgPtr, out);
        checkStatus(status, fallbackMessage: 'flm_manager_create failed');
        managerHandle = out.value;
      } finally {
        calloc.free(out);
      }
    });

    final lifecycleBridge = LifecycleBridge.attach(managerHandle);
    return FoundryLocal._(
      config: resolvedConfig,
      managerHandle: managerHandle,
      lifecycleBridge: lifecycleBridge,
    );
  }

  /// Live device profile. Recomputed each time — thermal state and available
  /// memory change quickly on a phone.
  DeviceProfile get deviceProfile {
    _ensureAlive();
    final bindings = NativeLibrary.instance.bindings;
    final out = calloc<Pointer<Char>>();
    try {
      final status =
          bindings.flm_manager_get_device_profile_json(_managerHandle, out);
      checkStatus(status,
          fallbackMessage: 'flm_manager_get_device_profile_json failed');
      try {
        final s = takeOutString(out.value);
        final json = jsonDecode(s) as Map<String, Object?>;
        return DeviceProfile.fromJson(json);
      } finally {
        bindings.flm_string_free(out.value);
      }
    } finally {
      calloc.free(out);
    }
  }

  /// Adjust mutable runtime settings without recreating the manager.
  void updateSettings(RuntimeSettings settings) {
    _ensureAlive();
    final bindings = NativeLibrary.instance.bindings;
    withCString(settings.toJsonString(), (ptr) {
      final status = bindings.flm_manager_update_settings(_managerHandle, ptr);
      checkStatus(status,
          fallbackMessage: 'flm_manager_update_settings failed');
    });
  }

  /// Forward an OS lifecycle event to the core. Bindings wire this to the
  /// platform's own notifications automatically via [LifecycleBridge]; apps
  /// rarely need to call it directly.
  void notifyLifecycle(FlmLifecycleEventKind event) {
    _ensureAlive();
    NativeLibrary.instance.bindings
        .flm_manager_notify_lifecycle(_managerHandle, event.code);
  }

  /// Validate, load and register a model from a local filesystem path.
  ///
  /// The returned [Model] is already loaded and ready for session creation.
  Future<Model> loadModel(
    String path, {
    String? executionProvider,
    Map<String, String>? providerOptions,
    void Function(Progress)? onProgress,
  }) async {
    _ensureAlive();
    final bindings = NativeLibrary.instance.bindings;
    final options = LoadOptions(
      executionProvider: executionProvider,
      providerOptions: providerOptions,
    );
    final optionsMap = options.toJson();
    final optionsJson = optionsMap.isEmpty ? null : jsonEncode(optionsMap);

    Sink<Progress>? sink;
    StreamController<Progress>? controller;
    if (onProgress != null) {
      controller = StreamController<Progress>();
      sink = controller.sink;
      controller.stream.listen(onProgress);
    }

    try {
      final result = await withCString<Future<Map<String, Object?>>>(
        path,
        (pathPtr) => withNullableCString<Future<Map<String, Object?>>>(
          optionsJson,
          (optionsPtr) => runProgressJob(
            abiCall: (progressPtr, completionPtr, userDataPtr, outJob) =>
                bindings.flm_manager_load_model_async(
              _managerHandle,
              pathPtr,
              optionsPtr,
              progressPtr,
              completionPtr,
              userDataPtr,
              outJob,
            ),
            onProgress: sink,
          ),
        ),
      );

      final handle =
          (result['model_handle'] as num?)?.toInt() ?? raw.FLM_INVALID_HANDLE;
      if (handle == raw.FLM_INVALID_HANDLE) {
        throw buildException(
          raw.FlmStatus.invalidHandle,
          message:
              'No model handle returned for path "$path" from flm_manager_load_model_async.',
          detail: result,
        );
      }
      return Model.fromHandle(handle);
    } finally {
      await controller?.close();
    }
  }

  /// Shut down the manager, cancel in-flight jobs and unload every model.
  ///
  /// After this returns you may still receive queued completion callbacks
  /// (with `FLM_ERROR_CANCELLED`) before [dispose] frees the underlying
  /// resources. Prefer [dispose], which chains shutdown into cleanup.
  void shutdown() {
    _ensureAlive();
    NativeLibrary.instance.bindings.flm_manager_shutdown(_managerHandle);
  }

  /// Release every native resource this manager owns. Idempotent.
  Future<void> dispose() async {
    if (_disposed) return;
    _disposed = true;
    final bindings = NativeLibrary.instance.bindings;
    bindings.flm_manager_shutdown(_managerHandle);
    bindings.flm_manager_release(_managerHandle);
    await _lifecycleBridge.detach();
  }

  void _ensureAlive() {
    if (_disposed) {
      throw StateError('FoundryLocal has been disposed.');
    }
  }

  /// Internal accessor for the raw manager handle. Only used by sibling
  /// classes in this package.
  @visibleForTesting
  int get managerHandle => _managerHandle;
}

/// Enumeration of lifecycle events. Mirrors the C `flm_lifecycle_event`.
enum FlmLifecycleEventKind {
  foreground(raw.FlmLifecycleEvent.foreground),
  background(raw.FlmLifecycleEvent.background),
  memoryWarning(raw.FlmLifecycleEvent.memoryWarning),
  memoryCritical(raw.FlmLifecycleEvent.memoryCritical),
  lowPower(raw.FlmLifecycleEvent.lowPower),
  thermalThrottling(raw.FlmLifecycleEvent.thermalThrottling);

  const FlmLifecycleEventKind(this.code);
  final int code;
}

File: bindings/ios/Sources/FoundryLocal/FoundryLocal.swift
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import Foundation
import FoundryLocalMobile

/// Root object of the Swift SDK. Wraps a single `flm_manager` handle.
///
/// Owns:
///   * The manager handle (released on ``close()``).
///   * A ``LifecycleObserver`` forwarding OS notifications to the core.
///
/// Not typically instantiated more than once per process. iOS supports several
/// simultaneously, but they'll contend for the model cache directory unless each is
/// configured with a distinct ``FoundryLocalConfig/appDataDir``.
public final class FoundryLocal: @unchecked Sendable {
    public let handle: flm_manager

    private let released = ManagedAtomicBool()
    #if canImport(Darwin)
    private let lifecycle: LifecycleObserver
    #endif

    // MARK: - Construction

    /// The initializer is synchronous — the core does its own work on a job thread —
    /// but it *may* block briefly while the runtime is bound. Prefer calling it once
    /// at app launch off the main thread.
    public convenience init(config: FoundryLocalConfig) throws {
        let json = try config.encodeAsJSON()
        var handle: flm_manager = 0
        let status = json.withCString { flm_manager_create($0, &handle) }
        if status != FLM_OK {
            throw FoundryLocalError.fromCurrent(status: status)
        }
        try self.init(handle: handle)
    }

    public init(handle: flm_manager) throws {
        guard handle != 0 else {
            throw FoundryLocalError(code: .invalidHandle, message: "manager handle must not be zero")
        }
        self.handle = handle
        #if canImport(Darwin)
        self.lifecycle = LifecycleObserver(manager: handle)
        #endif
    }

    deinit {
        close()
    }

    // MARK: - Lifecycle

    /// Shut down and release the manager. Cancels in-flight jobs, unloads models and
    /// blocks briefly on outstanding work — safe to call from any thread but not from
    /// the main queue when a large model is loaded.
    public func close() {
        if released.exchange(true) { return }
        #if canImport(Darwin)
        lifecycle.stop()
        #endif
        _ = flm_manager_shutdown(handle)
        _ = flm_manager_release(handle)
    }

    // MARK: - Load model from path

    /// Load a model directly from a local directory path and return a ready-to-use
    /// ``Model``.
    ///
    /// This is the recommended entry point for on-device inference. It validates the
    /// directory, loads the model through the native runtime, and returns a ready
    /// ``Model`` handle in one call.
    ///
    /// ```swift
    /// let model = try await sdk.loadModel(
    ///     at: "/path/to/model/dir",
    ///     executionProvider: "CoreML"
    /// )
    /// let chat = try model.createChatSession()
    /// ```
    ///
    /// - Parameter path: Absolute filesystem path to the model directory. The
    ///   directory must already exist on disk.
    /// - Parameter executionProvider: Optional execution provider override, e.g.
    ///   `"CoreML"`.
    /// - Parameter providerOptions: Optional key-value EP configuration, forwarded
    ///   as `provider_options` to the OGA session (e.g. `["use_fp16": "1"]`).
    /// - Parameter progress: Optional progress callback for the load phase.
    /// - Returns: A loaded ``Model`` ready for session creation.
    /// - Throws: ``FoundryLocalError`` if the path is invalid or loading fails.
    public func loadModel(
        at path: String,
        executionProvider: String? = nil,
        providerOptions: [String: String]? = nil,
        progress: ((Progress) -> Void)? = nil
    ) async throws -> Model {
        let progressHandler: (@Sendable (Progress) -> Void)? = progress.map { callback in
            let box = ProgressCallbackBox(callback)
            return { value in box.callback(value) }
        }
        let options = loadModelOptions(
            executionProvider: executionProvider,
            providerOptions: providerOptions
        )
        let payload: LoadModelPayload = try await runAsyncJob(
            progress: progressHandler,
            decode: { job in
                let text = try takeJobResultJSON(job)
                return try flmJSONDecoder.decode(LoadModelPayload.self, from: Data(text.utf8))
            },
            submit: { [handle] userData, onProgress, onComplete, outJob in
                path.withCString { pathPtr in
                    options.withCString { optionsPtr in
                        flm_manager_load_model_async(handle, pathPtr, optionsPtr, onProgress, onComplete, userData, outJob)
                    }
                }
            }
        )
        guard payload.modelHandle != 0 else {
            throw FoundryLocalError(
                code: .invalidHandle,
                message: "No model handle returned for path '\(path)'."
            )
        }
        return Model(handle: flm_model(payload.modelHandle))
    }

    // MARK: - Introspection

    /// Read the device profile the core uses for model placement.
    public func deviceProfile() throws -> DeviceProfile {
        let json = try readJSON { flm_manager_get_device_profile_json(handle, $0) }
        return try flmJSONDecoder.decode(DeviceProfile.self, from: Data(json.utf8))
    }

    /// Runtime settings that can be updated after creation. Fields left `nil` are
    /// preserved. Corresponds to the subset of ``FoundryLocalConfig`` accepted by
    /// `flm_manager_update_settings`.
    public func updateSettings(
        logLevel: FoundryLocalConfig.LogLevel? = nil,
        autoUnloadOnBackground: Bool? = nil
    ) throws {
        var payload: [String: Any] = [:]
        if let v = logLevel { payload["log_level"] = v.rawValue }
        if let v = autoUnloadOnBackground { payload["auto_unload_on_background"] = v }
        guard !payload.isEmpty else { return }
        let json = payload.jsonString() ?? "{}"
        let status = json.withCString { flm_manager_update_settings(handle, $0) }
        if status != FLM_OK {
            throw FoundryLocalError.fromCurrent(status: status)
        }
    }

    /// Manually push a lifecycle event, for cases where the standard notifications
    /// aren't sufficient (e.g. an app-internal thermal signal).
    public func notify(lifecycle event: LifecycleEvent) {
        _ = flm_manager_notify_lifecycle(handle, event.cValue)
    }

    // MARK: - Static ABI helpers

    /// Version of this Swift SDK, sourced from the core.
    public static var version: String {
        guard let ptr = flm_version_string() else { return "unknown" }
        return String(cString: ptr)
    }

    /// Version of the underlying ONNX Runtime GenAI runtime, or `nil` when the runtime
    /// is not linked in — usually a build error rather than a runtime concern.
    public static var runtimeVersion: String? {
        guard let ptr = flm_runtime_version_string() else { return nil }
        return String(cString: ptr)
    }

    /// Whether the runtime is present and loadable at all. Fails silently — never
    /// logs, never crashes — so it is safe to call before any config exists.
    public static var isRuntimeAvailable: Bool {
        flm_is_runtime_available() != 0
    }

    /// Set the minimum severity forwarded through ``FoundryLocal/logSink``.
    public static func setLogLevel(_ level: FoundryLocalConfig.LogLevel) throws {
        let cLevel: flm_log_level
        switch level {
        case .verbose: cLevel = FLM_LOG_VERBOSE
        case .debug: cLevel = FLM_LOG_DEBUG
        case .info: cLevel = FLM_LOG_INFO
        case .warning: cLevel = FLM_LOG_WARNING
        case .error: cLevel = FLM_LOG_ERROR
        case .fatal: cLevel = FLM_LOG_FATAL
        case .off: cLevel = FLM_LOG_OFF
        }
        let status = flm_set_log_level(cLevel)
        if status != FLM_OK {
            throw FoundryLocalError.fromCurrent(status: status)
        }
    }

    private func loadModelOptions(
        executionProvider: String?,
        providerOptions: [String: String]?
    ) -> String {
        var payload: [String: Any] = [:]
        if let executionProvider { payload["execution_provider"] = executionProvider }
        if let providerOptions, !providerOptions.isEmpty { payload["provider_options"] = providerOptions }
        return payload.jsonString() ?? "{}"
    }
}

private final class ProgressCallbackBox: @unchecked Sendable {
    let callback: (Progress) -> Void

    init(_ callback: @escaping (Progress) -> Void) {
        self.callback = callback
    }
}

// MARK: - Lifecycle event mirror

public enum LifecycleEvent: Sendable {
    case foreground
    case background
    case memoryWarning
    case memoryCritical
    case lowPower
    case thermalThrottling

    var cValue: flm_lifecycle_event {
        switch self {
        case .foreground: return FLM_LIFECYCLE_FOREGROUND
        case .background: return FLM_LIFECYCLE_BACKGROUND
        case .memoryWarning: return FLM_LIFECYCLE_MEMORY_WARNING
        case .memoryCritical: return FLM_LIFECYCLE_MEMORY_CRITICAL
        case .lowPower: return FLM_LIFECYCLE_LOW_POWER
        case .thermalThrottling: return FLM_LIFECYCLE_THERMAL_THROTTLING
        }
    }
}
